### PR TITLE
Search questions dev

### PR DESCRIPTION
**Pull request by jensptan**
Search tables of questions by tag on the Question Index page and the Question Set Edit and New pages. Tables of questions are populated with questions tagged with courses the professor or TA is registered as an ASSISTANT for. Live searching filters table based on questions' tags.

### CODE REVIEW

File: app/models/registration.rb [hunk 1]
```diff
@@ -2,5 +2,5 @@ class Registration < ApplicationRecord
   belongs_to :user
   belongs_to :course
 
-  validates :role, inclusion: { in: %w(STUDENT ASSISTANT) }
+  validates :role, inclusion: { in: %w(STUDENT ASSISTANT PROFESSOR) }
```
Comment (jt1796) [inline app/models/registration.rb]: Why `PROFESSOR`?

Comment (jensptan) [inline app/models/registration.rb]: That was from before "ASSISTANT" was clarified as being either professor or TA. Allow me to remove that real quick.

File: app/views/question_sets/_form.html.erb [hunk 2]
```diff
@@ -26,7 +26,62 @@
...
+    <!-- current_user's courses do not have any questions -->
+    <p>There are no questions associated with the courses you teach. Please create a question to add to the question set.</p>
+  <% end %>
+
   <div class="actions">
     <%= form.submit %>
   </div>
 <% end %>
+
+<script>
```
Comment (jt1796) [inline app/views/question_sets/_form.html.erb]: One suggestion I have is to put this script in a file in `app/assets/javascripts` (maybe `question_search.js`) and `qSetQTagSearch()` will automatically be available for use. Then it won't have to be in two different pages

**Comment by jt1796** [comment]
Pulled this locally and everything seems to work great! :rocket: 
Just one suggestion about the js and I think this is good to go

File: app/views/question_sets/_form.html.erb (hunk 2 above)
Comment (jensptan) [inline app/views/question_sets/_form.html.erb]: The methods are a bit different, though, based on which column in the table the tags are in (1 or 2).

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  get "/question_sets_new" => 'question_sets#new'
```
Comment (jt1796) [inline config/routes.rb]: Is this needed?

Comment (jensptan) [inline config/routes.rb]: No, it is not. Removing it now.

File: app/views/question_sets/_form.html.erb [hunk 4]
```diff
@@ -26,7 +26,62 @@
...
+    <table id="questionTagTable" border="1">
+      <tr>
+        <th></th>
+        <th style="min-width:250px">Body</th>
+        <th style="min-width:100px">Tags</th>
+      </tr>
+      <!-- loop thru questions from controller
+            (questions associated with courses the current_user is
+            a professor or assistant for) -->
+      <% @questions.each do |q| %>
```
Comment (jt1796) [inline app/views/question_sets/_form.html.erb]: Almost forgot one last thing, I think this table is supposed to replace this: https://github.com/jds2017/ALRails/blob/59a0873f52b9e90e0f64126fb727a74b0e2ac04e/app/views/question_sets/_form.html.erb#L19

I think you can delete that^ but then you'll also need to copy the logic for the checkbox. I also noticed that when I tried to save something with nothing checked, it errored out

File: app/views/question_sets/_form.html.erb (hunk 2 above)
Comment (jt1796) [inline app/views/question_sets/_form.html.erb]: You could change the function just slightly to read the last column of the table, that way it will work on either the one with the checkboxes or without.

Something like:
```
var number_columns =  document.getElementById("questionTagTable").rows[0].cells.length
for (i = 0; i < tr.length; i++) {
  //get tags column for this row
  td = tr[i].getElementsByTagName("td")[number_columns - 1];
```

might work

File: app/views/question_sets/_form.html.erb (hunk 4 above)
Comment (jensptan) [inline app/views/question_sets/_form.html.erb]: I have removed the duplicate table.

I tried saving a question set with nothing checked on the current master branch, and it errored out as well. That issue is not related to these changes.

File: config/routes.rb (hunk 3 above)
Comment (jensptan) [inline config/routes.rb]: Good catch, removed it

**Review by jt1796 — APPROVED** [review]
I'm fine with leaving the no questions error as a known bug that can be fixed after the demo.
